### ISSUE TITLE
cerebroApp

### ISSUE THREAD

**Issue by romanhaa** [Bioconductor/Contributions#1355]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/romanhaa/cerebroApp

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by romanhaa** [comment]
I already sent an email to the Bioc-devel mail list regarding the version, which causes a warning because it is currently at v1.1.2. I made the mistake of moving to v1.0 (and v1.1) too early. My idea was to release the package on Bioconductor as v1.2 and kept it at 1.1.2 to have some room to bump the version as part of the revision process. Normally, I wouldn't see any problem in decreasing the version below v1.0 and then starting on Bioconductor with v1.0. However, considering that I have already released standalone version of Cerebro (https://github.com/romanhaa/Cerebro), I hope that we can find a solution that doesn't involve changing the standalone version. I'm open to ideas on how to handle this in an adequate way. Thank you!

**Comment by lawremi** [comment]
I would expect this to support a SingleCellExperiment object and other Bioc infrastructure.

**Comment by romanhaa** [comment]
I have heard this comment before (during the revision of the paper) and decided to rely on existing conversion functions for popular scRNA-seq data formats. For example, a SingleCellExperiment object can be converted into a Seurat object (to subsequently export it) using the `Seurat::Convert()` function. I made a scheme here: 
![image](https://user-images.githubusercontent.com/18103560/73554097-64a51d80-444b-11ea-8226-754fccd332e3.png)

Would this not be sufficient?

**Comment by lawremi** [comment]
No, Bioconductor packages should extend Bioconductor. Of course, you could do the conversion internally, but there shouldn't be any artificial barriers for Bioconductor users.

**Comment by romanhaa** [comment]
I would argue that making an alternative for an existing conversion function would create unnecessary redundancy. It would really just be

```r
new_conversion_function <- function(SCE_object) return(Seurat::Convert(SCE_object))
```

But if this is a major problem then I'll try to find a solution.

If, however, you expect every single preparation function (not just the export function) to be compatible with an SCE object, then this will require substantial work which I'm not sure I can do.

**Comment by lawremi** [comment]
At the very least, `launchCerebro()` should take a `SingleCellExperiment`. Maybe @LTLA has some thoughts.

**Comment by romanhaa** [comment]
Actually, `launchCerebro()` doesn't take any object at all. The data, after being exported to a file, is loaded through the Shiny interface. I thought about adding a parameter to `launchCerebro()` so that an object can be passed to it directly, which would be convenient, but it's not straightforward so I decided to put that aside for now.

**Comment by lawremi** [comment]
The goal of the Bioconductor project is to build an interoperable platform for genomics. Since this does not directly interoperate with or depend on Bioconductor, it does not seem like it belongs. I'll let others weigh in though.

**Comment by LTLA** [comment]
There are a few philosophical considerations here that touch on the purpose of the Bioconductor project and the user expectations regarding the behavior of various Bioconductor packages.

One of the cornerstones of Bioconductor is the expectation that all packages should work together so that together they are greater than the sum of their parts. This lofty ideal is underwritten by the use of the same data structures in the interfaces of various packages, e.g., `GRanges`, `SummarizedExperiment`, `DNAStringSet`, and so on. The result is that we can take the output of one Bioconductor package and plug it into another Bioconductor package without the need for class transformations. Deviations from this policy cause headaches for both users and developers; for example, **edgeR** defines its own classes for ancient historical reasons, and it is a headache to work around those classes when we want to use **edgeR** with or inside other Bioconductor packages. (And I'm an **edgeR** maintainer! So if it's annoying for me, one can only imagine how annoying it is for the general user population.) In your case, if I'm coming from **scran** or **scater** or any of the other ~70 Bioconductor packages that use a `SingleCellExperiment` object, and I want to use **cerebroApp**, I would have to know where the converter lives (in **Seurat**), know that `Convert()` is the function responsible for doing the conversion, and hope that their converter doesn't discard data aspects that I was interested in.

Another key aspect of Bioconductor is the build release cycle that guarantees a 6-month window of stability. Users can expect that packages in Bioconductor-release are stable in that the interface and output will not change within the release interval. This sounds boring but is very important; I don't want to wake up and realize that all my analysis scripts no longer work because the latest version of a particular package (and I won't name any names here) has changed its data structure and the names of its functions. (Also, no, containers are not a real solution for most of our target audience.) If your package relies on other Bioconductor packages, you can rest assured that you won't be broken in release provided that those other packages respect the stability expectation during the release window. As soon as your package depends on CRAN packages, you are relying on the goodwill of CRAN maintainers to not introduce breaking changes. Personally, the only non-base/recommended CRAN packages I trust unconditionally are as follows:

- **Rcpp**
- **shiny**

And even then, there have been some hairy moments that caused me to lose sleep patching packages in release. (Specifically, a version bump was required when **Rcpp** updated its ABI in version 1.0.0.) For all other CRAN packages, I aim to limit the exposure of my packages to them, thus mitigating the repair work required upon breakage; or I write abstraction layers such as **BiocSingular** and **BiocNeighbors** so that I only have to fix things in one place rather than patching 10 different packages at 1 am in the morning. Now, I can't possibly imagine **Seurat** changing their data structure between versions, but if they did, your package would effectively be crippled - in release! - until you managed to find some time to fix it. And if your package is popular (and let's hope it is) then you will get complaints, lots of them, sometimes angry ones. Yeah, some people are like that, and it's not fun.

Perhaps a personal anecdote would be illustrative. I once submitted a package for analyzing Hi-C data where I was using my own custom (non-Bioconductor) data structures. The reviewer suggested that I use some more standard objects, but I ignored him - I had a manuscript to submit, after all, and who cares about these technical issues anyway? We're doing science! Well, it turns out that people _need_ to use my package alongside other packages, because my package doesn't do everything - in fact, it only does one thing - so I _really needed_ interoperability. It took two years to repair the damage and make my package work with other Bioconductor packages; this was because once I released my package, I had to respect the existing interface, so it took several release cycles supporting both the new and old interfaces before I was able to migrate. Think about it: two years because I kicked the can down the road for a week's worth of work, that's a pretty bad deal.

So to sum up: if you submit a package to Bioconductor, there is an expectation that your package can work seamlessly with other Bioconductor packages, and your implementation should support that. The safest and easiest way to do that is to use Bioconductor data structures, with the minimum standard of being able to accept such objects even if you do not rely on them internally (though you should). If you don't want to do that, then that's a shame, but it would suggest that Bioconductor would not be the right place to host this package.

**Comment by romanhaa** [comment]
Thank you @LTLA for sharing your thoughts and experience, I really appreciate it! You made some very good points.

I'd like to explain/comment a few things:

1) When considering to submit my package to Bioconductor, I clearly underestimated the importance of interoperability with 'Bioconductor' data structures, such as `SingleCellExperiment` in this case. Also, I was asked by both reviewers and users to add cerebroApp to Bioconductor. When looking at the Bioconductor website, the first sentence I read was the following, and that seemed very much in line with the goal of my package:

> Bioconductor provides tools for the analysis and comprehension of high-throughput genomic data.

2) Due to the design philosophy of Cerebro, which is to not create another data structure to collect the data for visualization and instead allow the user to extend their existing data structure (e.g. a Seurat object), supporting another data structure multiplies the effort almost linearly with every supported format<sup>1</sup>. The choice of going with a Seurat object was very practical for two reasons: First, in my department and among my more distant colleagues, the Seurat framework is by far the most commonly used. To give most people access to Cerebro, I therefore decided to start with that. Second, I think it's quite beginner-friendly and provides plenty of conversion functions, even though they don't always transfer all the information. Philosophically, I very much like the idea of a common data structure across different scRNA-seq frameworks, such as the `SingleCellExperiment` object, to facilitate conversion and increase compatibility. With Seurat being such a popular framework, it is sadly a bit difficult to apply it in practice.

3) The primary element of Cerebro/cerebroApp was always the Shiny interface. All functions that actually do some analysis and generate results, i.e. all functions except `exportFromSeurat()` and `launchCerebro()`, are considered optional and, in the beginning, there were only very few of them. However, as they increased in number, they started to significantly enhance the potential of Cerebro. This brings me back to the previous point where I would run into problems when holding this data in a separate object.

4) The Seurat package actually has changed their object structure in the past (moving from v2.3 to v3.0) to allow storing different assays (e.g RNA + ATAC) in the same object. The idea of that changing again is quite scary, but I consider it stable now.

Considering @LTLA's points, it seems that the cerebroApp is either not ready to be added to Bioconducter yet, or might never be due to incompatibilities in design choices. This is a bit sad since I would've liked to give users the benefits it would've had, and I must admit I was a bit disappointed at first, but it might indeed be for the best to wait and reconsider, and perhaps make the integration with Bioconductor smoother.

Finally, if I may ask for an advice from you, considering that cerebroApp depends on both CRAN and Bioconductor packages, would you suggest adding it to CRAN or leave it on GitHub?

<sup>1</sup> I'd be very happy to receive feedback about the choice of not creating a new data structure since I'm effectively the only developer, and not one with much experience. I thought it was a good idea because scRNA-seq analysis is often an iterative process. By extending the existing object, users don't have to re-generate results in a hypothetical Cerebro object every time they want to export the data for visualization. Thereby, the data always stays up-to-date. However, this evidently has some drawbacks.

**Comment by hpages** [comment]
> considering that cerebroApp depends on both CRAN and Bioconductor packages, would you suggest adding it to CRAN or leave it on GitHub?

Even though it's technically possible to put it on CRAN -- some people do it and CRAN seems to have nothing against hosting packages that depend on Bioconductor -- it is not an ideal situation. That's because Bioconductor release cycles have the potential to break CRAN packages that rely on them every 6 month. This doesn't happen if the package is part of Bioconductor ;-)

That being said it looks like cerebroApp only depends on Bioconductor packages biomaRt, ggtree, and GSVA. So the question is how likely is it that these packages will make changes that break cerebroApp. I would say quite unlikely but you never know.

Note that leaving cerebroApp on GitHub doesn't really avoid the problem either. There is probably a way to make sure that the package stays in sync with the current release version of Bioconductor (by properly defining remotes and setting up CI) but that wouldn't necessarily prevent breakage every 6 months when a new version of Bioconductor is released. If you choose to stay in sync with Bioc-devel (via proper remotes and CI setup) then you have a different problem which is that now your package might stop working for people using stable R + BioC release (that would typically be more than 90% of your users).

Since we are on the topics of dependencies, cerebroApp imports 29 packages which is A LOT (after `library(cerebroApp)`, `sessionInfo()` reports  99 packages "loaded via a namespace"). Are you sure you need all of them? The more deps the more fragile a package is. At least it seems that maybe some of them could be moved to Suggests. For example you only use biomaRt in your `getMarkerGenes()` function to retrieve GO terms as a "side task" (i.e. failure to retrieve the terms doesn't prevent your function from performing its main task). That means your could suggest biomaRt instead of importing it. I would recommend you try to reduce your deps and move to Suggests those that are not essential to your package.

**Comment by lawremi** [comment]
Strongly recommend CRAN over GitHub. CRAN gives guarantees that your package passes tests in conjunction with the rest of the CRAN repository. Thus, users can trust that if they install your package, it will work with everything else they got from CRAN. Also, the user can use tools built into R to install from CRAN, and it's easier to provide a stable (released) version of the package via CRAN.

**Comment by LTLA** [comment]
> supporting another data structure multiplies the effort almost linearly with every supported format

For comparison, many of the **scran** and **scater** S4 generics will support both raw matrix input and a `SingleCellExperiment` (or even a `SummarizedExperiment`) input. This is trivial to achieve with S4 dispatch and allows people to use those packages even if they never want to touch a SE.

For further context, you should have a look at examples of multiple dispatch, e.g., `findKNN` from **BiocNeighbors**, matrix subsetting (i.e. `[`) in **Matrix**. Here, the number of required methods increases quadratically with the number of input types that need to be supported. Linear is nothing.

> * The Seurat package actually has changed their object structure in the past (moving from v2.3 to v3.0) to allow storing different assays (e.g RNA + ATAC) in the same object. The idea of that changing again is quite scary, but I consider it stable now.

Yes, I know, I was being sarcastic. All I will say is that the best way to figure out how people will behave in the future is to see how they have behaved in the past.

> Finally, if I may ask for an advice from you, considering that cerebroApp depends on both CRAN and Bioconductor packages, would you suggest adding it to CRAN or leave it on GitHub?

Or just add support for SCE inputs and submit it to Bioconductor. See #1302 for an example of a package that has support for both types of objects. If you only have two essential functions, this should be trivial.

**Comment by romanhaa** [comment]
Thank you all @hpages @lawremi @LTLA for the suggestions.

> Since we are on the topics of dependencies, cerebroApp imports 29 packages which is A LOT (after library(cerebroApp), sessionInfo() reports 99 packages "loaded via a namespace"). Are you sure you need all of them?

I guess I could rewrite some parts and move some packages to `Suggests` in order to reduce the number of dependencies. Unfortunately, the Shiny interface needs a few packages that themselves have multiple dependencies. And the fact that cerebroApp acts at the end of the scRNA-seq workflow doesn't help either. But I'll make sure to have a look at this.

> Or just add support for SCE inputs and submit it to Bioconductor. See #1302 for an example of a package that has support for both types of objects. If you only have two essential functions, this should be trivial.

At the moment this would be my preferred choice since I still think that Bioconductor would be the best place for cerebroApp, assuming I'll meet the community guidelines explained before. I'll probably just give myself some more time, implement proper support for SCE objects, and aim to re-submit for the next release cycle. I won't be able to do it by April.

> Strongly recommend CRAN over GitHub.

I'll see how the implementation of support for SCE objects goes. If it goes well, I'll re-submit to Bioconductor. If not, I'll go for CRAN. I'm not planning to leave it on GitHub forever.

Should I close this and re-open a new issue when I'm ready or leave it open? As I said, this will take months as I have a thesis to write :/

**Comment by mtmorgan** [comment]
I'll close the issue; if you choose to submit in the future please add a comment to the closed issue and tag @mtmorgan that you'd like to reopen.
